### PR TITLE
bugfix/13868-hovering-pie-in-stock 

### LINKED ISSUE #13868: Pie series in Highstock - Error when hovering
#### Expected behaviour
Hovering over the pie series without errors.

#### Actual behaviour
Error when hovering over the pie series: `Uncaught TypeError: Cannot read property 'lineWidth' of undefined`

#### Live demo with steps to reproduce
http://jsfiddle.net/BlackLabel/fLxbv0pk/

#### Product version
Highstock v8.1.2

#### Affected browser(s)
All

### LINKED ISSUE #16085: Pie in stock with accessibility- error in the console caused by navigator
#### Expected behaviour
No errors in the console.

#### Actual behaviour
When the accessibility module is loaded, hovering over the pie series with `stockChart` constructor generates an error in the console.
Without a navigator works fine.

#### Live demo with steps to reproduce
http://jsfiddle.net/BlackLabel/yx54bkq9/

Without navigator: http://jsfiddle.net/BlackLabel/f1zm87hc/

#### Product version
Highstock JS v9.1.2 (2021-06-16) on branch `bugfix/13868-hovering-pie-in-stock` which fixes some issues 

#### Affected browser(s)
All

### PR DESCRIPTION
**Pull request by karolkolodziej**
Fixed #13868, #16085, hovering over a `pie` series in a stock chart generated errors in the console.

Closes #16085.

### CODE REVIEW

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks!

We need a more general solution, as this bug is not specific to stock charts: https://jsfiddle.net/highcharts/pbj5ruqx/

The pie series class (along with some other classes) has a property called `noSharedTooltip` to prevent similar cases like this for shared tooltip. Split tooltips are similar to shared in the sense that they are grouped by the X axis value. So I would look for a solution around the `noSharedTooltip` property.

File: ts/Core/Tooltip.ts [hunk 1]
```diff
@@ -1379,6 +1380,12 @@ class Tooltip {
 
         // single point tooltip
         } else {
+            // Hovering over series that tooltip shouldn't be split, #13868.
+            if (!isArray(pointOrPoints) && pointOrPoints.series.noSplitTooltip) {
+                tooltip.options.split = false;
+                tooltip.split = false;
+            }
```
Comment (pawelfus) [inline ts/Core/Tooltip.ts:1387]: I'm afraid this might not work, because there's no reset to `split=true`. I wanted to try this demo: https://jsfiddle.net/BlackLabel/5wdz4a8y/1/ - but it fails

**Review by pawelfus — CHANGES_REQUESTED** [review]

File: ts/Core/Tooltip.ts [hunk 2]
```diff
@@ -69,6 +69,7 @@ declare module './Series/PointLike' {
 declare module './Series/SeriesLike' {
     interface SeriesLike {
         noSharedTooltip?: boolean;
+        noSplitTooltip?: boolean;
```
Comment (TorsteinHonsi) [inline ts/Core/Tooltip.ts:72]: Probably better to check `noSharedTooltip` for split tooltip too. A split tooltip is also a form of shared tooltip, in the sense that multiple series are shown together. Also, `noSharedTooltip` is not limited to pie series.

**Review by pawellysy — CHANGES_REQUESTED** [review]
Awaiting working version

File: ts/Core/Tooltip.ts (hunk 2 above)
Comment (karolkolodziej) [reply ts/Core/Tooltip.ts:72]: I decided to use `noSplitTooltip` because series like for example `scatter` might have splitted tooltip, when shared doesn't work for them.

File: ts/Series/PackedBubble/PackedBubbleSeries.ts [hunk 3]
```diff
@@ -1473,7 +1473,6 @@ interface PackedBubbleSeries {
     forces: Array<string>;
     hasDraggableNodes: boolean;
     isCartesian: boolean;
-    noSharedTooltip: boolean;
```
Comment (pawelfus) [inline ts/Series/PackedBubble/PackedBubbleSeries.ts:1476]: Q: Why removing props from the interface? By default `noSharedTooltip` is optional, but in this particular series it's always defined (and set to `true`). Probably for Booleans, it's not a big deal, but with other types, it might change (and require) more code.

Comment (karolkolodziej) [reply ts/Series/PackedBubble/PackedBubbleSeries.ts:1476]: I removed `noSharedTooltip` from the `PackedBubbleSeries` interface because this property has been already declared in the `Tooltip.ts` file as a `SeriesLike` interface.

Should it (`noSharedTooltip` property) be declared in the interface for each series individually?

@bre1470 could you please help us decide which pattern is better?

Comment (bre1470) [reply ts/Series/PackedBubble/PackedBubbleSeries.ts:1476]: If it is possible, that a series with shared tooltips inherits from a series without shared tooltips, it is better to define the property more generic. In that case `SeriesLike` in `Tooltip.ts` should be enough.

Comment (pawelfus) [reply ts/Series/PackedBubble/PackedBubbleSeries.ts:1476]: > Should it (noSharedTooltip property) be declared in the interface for each series individually?

Sorry, let me rephrase. Not for each series, but just the ones that overwrite the property. In other (coding) words:

```ts
interface SeriesLike {
  noSharedTooltip?: boolean
} 

class ColumnSeries extends SeriesLike {
}

class PieSeries extends ColumnSeries {
  noSharedTooltip: boolean // or even "true"
}

PieSeries.prototype.noSharedTooltip = true;

// Later in code:
// if (this.noSharedTooltip) { ... } <-- fine! 
```

As I said above, it's not a big deal with booleans, I just wanted to discuss this :) With e.g. objects it's a bit more troublesome:

```ts
interface customObjectInterface {
  x: number,
  y: number
}

interface SeriesLike {
  customObject?: customObjectInterface
} 

class ColumnSeries extends SeriesLike {
}

class PieSeries extends ColumnSeries {
  customObject: customObjectInterface
}

PieSeries.prototype.customObject = { x: 0, y: 0 };

// Later in code:
// `this` in an instance of PieSeries
if (this.customObject) {
   doSomething(
     this.customObject.x,
     this.customObject.y
   )
}
// Or
   doSomething(
     this.customObject?.x,
     this.customObject?.y
   )

// vs

doSomething(
  this.customObject.x,
  this.customObject.y
) // <-- `customObject` is defined
```

**Review by pawellysy — APPROVED** [review]

File: ts/Core/Series/Point.ts [hunk 4]
```diff
@@ -1670,7 +1670,7 @@ class Point {
                     }
                 }
 
-                if (!chart.styledMode && stateMarkerGraphic) {
+                if (!chart.styledMode && stateMarkerGraphic && point.state !== 'inactive') {
```
Comment (raf18seb) [inline ts/Core/Series/Point.ts:1673]: Could you shorten this to keep the 80 characters limit?

File: ts/Core/Tooltip.ts [hunk 5]
```diff
@@ -1392,8 +1396,25 @@ class Tooltip {
         if (text === false) {
             this.hide();
         } else {
+            // When previously split tooltip was rendered and now if
+            // hovering over another series where split tooltip should not
+            // be used, destroy the label in order to create ordinary one.
+            // #13868
+            if (tooltip.split && point.series.noSplitTooltip && tooltip.tt && tooltip.label) {
```
Comment (raf18seb) [inline ts/Core/Tooltip.ts:1403]: Could you shorten this to keep the 80 characters limit?

File: ts/Core/Tooltip.ts [hunk 6]
```diff
@@ -1392,8 +1396,25 @@ class Tooltip {
...
+                tooltip.label = tooltip.label.destroy();
+                tooltip.tt = tooltip.tt.destroy();
+                this.cleanSplit();
+            }
+            // When previously tooltip was generated for noSplitTooltip series
+            // and now split one should be rendered, destroy the label.
+            if (tooltip.split && !point.series.noSplitTooltip &&
+                !(tooltip.tt || !tooltip.options.headerFormat) && tooltip.label
+            ) {
```
Comment (raf18seb) [inline ts/Core/Tooltip.ts:1412]: More readable
```js
if (
    tooltip.split &&
    !point.series.noSplitTooltip &&
    !(tooltip.tt || !tooltip.options.headerFormat) &&
    tooltip.label
) {
...
```

File: samples/unit-tests/series-pie/pie/demo.js [hunk 7]
```diff
@@ -320,3 +320,41 @@ QUnit.test(
...
+    chart.tooltip.refresh(chart.series[0].points[0]);
+    chart.tooltip.refresh(chart.series[1].points[0]);
+    chart.tooltip.refresh(chart.series[0].points[0]);
+    assert.ok(
+        true,
+        `After hovering over two series where one of them should not have
+        a splitter tooltip, while tooltip split explicitly set to true,
+        there should be no errors in the console.`
+    );
+});
```
Comment (raf18seb) [inline samples/unit-tests/series-pie/pie/demo.js:360]: Could we make it one QUnit test instead of 2 separate? Using `chart.update()` for example. It could improve the performance of the tests.

File: samples/unit-tests/series-pie/pie/demo.js [hunk 8]
```diff
@@ -320,3 +320,41 @@ QUnit.test(
         );
     }
 );
+
+QUnit.test('Pie chart initialized through the stockChart constructor, (#14773).', function (assert) {
```
Comment (raf18seb) [inline samples/unit-tests/series-pie/pie/demo.js:324]: Using arrow function would shorten this line a little

**Review by raf18seb — CHANGES_REQUESTED** [review]

File: samples/unit-tests/series-pie/pie/demo.js (hunk 7 above)
Comment (karolkolodziej) [reply samples/unit-tests/series-pie/pie/demo.js:360]: Made it but found one issue while testing in karma #16085

**Review by raf18seb — APPROVED** [review]
Thanks!

**Review by pawelfus — APPROVED** [review]

**Comment by khlieng** [comment]
@oysteinmoseng do internal series need forced markers?

**Comment by oysteinmoseng** [comment]
@khlieng If you are referring to the navigator, then yes

File: ts/Core/Tooltip.ts [hunk 9]
```diff
@@ -1392,8 +1396,31 @@ class Tooltip {
...
+            // When previously tooltip was generated for noSplitTooltip series
+            // and now split one should be rendered, destroy the label.
+            if (tooltip.split &&
+                !point.series.noSplitTooltip &&
+                !(tooltip.tt || !tooltip.options.headerFormat) &&
+                tooltip.label
+            ) {
+                tooltip.label = tooltip.label.destroy();
+            }
+
```
[the remark this answers is not in the document]
Comment (karolkolodziej) [reply ts/Core/Tooltip.ts]: I've managed to move it to `getLabel` and slightly simplify it.
Checking for that class covers some edge cases here and there which I didn't consider so thanks for that hint ;)

File: ts/Core/Tooltip.ts (hunk 2 above)
Comment (karolkolodziej) [reply ts/Core/Tooltip.ts:74]: Unfortunately, it is the case for the flag series.
I'm afraid we have to have separate properties for that.

Comment (TorsteinHonsi) [reply ts/Core/Tooltip.ts:74]: As I recall it was never the intention that flags should have a split tooltip. I think it just happened and has been patched after. I'm pushing a commit with some simplifications.

**Comment by TorsteinHonsi** [comment]
@karolkolodziej I committed som simplifications, specifically removed `noSplitTooltip` because of the above discussion. I don't think it was intentional that flag series should have a split tooltip. So now we're going back to the single tooltip as is was before `split` was introduced.

Also, I simplified the destruction bit to simply run `this.destroy()`. Can you see if I am missing something there? 

https://github.com/highcharts/highcharts/blob/de1a1a699c1687da99097602b0d02091c4e559ed/ts/Core/Tooltip.ts#L493-L498

**Comment by karolkolodziej** [comment]
It's fine for me if you are ok with the flag having a different tooltip ;)

The only thing I would change is the comment above that if statement describing why we are destroying it.
```
When previously split tooltip was rendered and now 
if hovering over another series where split tooltip should not be used
(or the other way around), destroy it to create a proper one
for the currently hovering series, #13868.
```

**Review by TorsteinHonsi — APPROVED** [review]
